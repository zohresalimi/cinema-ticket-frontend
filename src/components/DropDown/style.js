import styled from "styled-components";

export const Wrapper = styled.div`
  display: inline;
  box-sizing: border-box;
  display: flex;
  justify-content: flex-end;
`;

export const MenuLabel = styled.ul`
  color: #fff;
  display: flex;
  flex-direction: column;
  position: absolute;
  top: 50px;
  right: 16px;
  z-index: 9;
  list-style: none;
  align-items: flex-end;
  text-align: center;
`;

export const ItemList = styled.li`
  font-family: "sfMedium";
  padding: 0.25em 2em 0.25em 2em;
  width: 100%;
  color: #000;
  background: white;
  line-height: 30px;
  padding: 0.25em 2em 0.25em 2em;
  cursor: defaul;
  border-bottom: 1px solid #eee;
  transition: all 0.25s ease;
  &:hover,
  &.selected {
    background: #f7f7f7;
    color: #4a4a4a;
  }
`;

export const Button = styled.button`
  display: inline-flex;
  align-items: center;
  padding: 0px 30px;
  font-size: 14px;
  justify-content: center;
  position: relative;
  font-size: 20px;
  color: #fff;
  background-color: Transparent;
  background-repeat: no-repeat;
  border: none;
  cursor: pointer;
  overflow: hidden;
  outline: none;
  appearance: none;
  box-shadow: none;
  transition: ease-in-out 0.2s;
  & hover: {
    color: green;
  }
  &:focus {
    outline: none;
  }
`;

export const CaratContainer = styled.div`
  transform: ${(props) => (props.isOpen ? "rotate(180deg)" : "rotate(0deg)")};
  transition: all 0.2s ease;
  width: 20px;
`;
