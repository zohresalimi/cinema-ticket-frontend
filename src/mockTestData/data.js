import React from "react";
import { Router } from "@reach/router";
import AppContext from "../store/context";
import Theme from "../Styles/Theme";

export const getTestStore = () => {
  return {
    dispatch: () => {},
    state: {
      user: {
        name: "",
        email: "",
      },
      movies: {
        children: [
          {
            _id: "5fa71823adc463397a50c501",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "trolls 2: world tour",
            originalTitle: "swedish",
            premiere: "2020-10-03T15:18:52.722Z",
            rooms: ["5fa653d0a7f1201999dd1ac0"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        premiered: [
          {
            _id: "5fa71823adc463397a50c502",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "follow me",
            originalTitle: "swedish",
            premiere: "2020-10-05T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        upcoming: [
          {
            _id: "5fa71823adc463397a50c503",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            coverImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/68bfeef55ada4e7ab0606db2af0f40ad.jpg?width=240&version=EA2E80EE161E5DA161C6E26487054C9F&format=webp",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            largeImage:
              "https://catalog.cinema-api.com/cf/images/ncg-images/79276cead82341d2a004f8e8f99a88e9.jpg?width=1920&version=08F0C83177D59FB94C4C809F63F67E2A&format=webp",
            name: "honest thief",
            originalTitle: "swedish",
            premiere: "2021-01-03T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
        ],
        allMovies: {
          "5fa71823adc463397a50c501": "children",
          "5fa71823adc463397a50c502": "premiered",
          "5fa71823adc463397a50c503": "upcoming",
        },
      },
      cinemas: [
        {
          _id: "5fa64fe234fbbf9945f3a3ce",
          name: "filmstaden kista",
          purchaseEndTime: "21:30",
          purchaseStartTime: "8:00",
          rooms: ["5fa653c0643c06586bba7d78", "5fa653d0a7f1201999dd1ac0"],
        },
      ],
      ticket: {
        movie: {},
        status: false,
        quantity: 0,
        price: 0,
        seatNumbers: new Map(),
        user: {},
        showing: {},
      },
      showings: [
        {
          _id: "5fbe9bd31f58462b5c759b5c",
          capacity: 4,
          cinema: "5fa64fe234fbbf9945f3a3ce",
          endTime: null,
          movie: {
            _id: "5fa71823adc463397a50c502",
            actors: ["Holland Roden", "Ronen Rubinstein", "Keegan Allen"],
            age: "7",
            description:
              "Poppy and Kvist embark on an adventure that will take them far beyond",
            director: "aaron woodley",
            duration: "1:31",
            genre: ["family", "children"],
            images: [],
            name: "follow me",
            originalTitle: "swedish",
            premiere: "2020-10-05T15:18:52.722Z",
            rooms: ["5fa653c0643c06586bba7d78"],
            trailerUrl: "https://youtu.be/P9A6hOg9QQ0",
          },
          price: 100,
          room: { _id: "5fa653c0643c06586bba7d78" },

          seats: [
            [
              { seatNumber: 1, taken: "available" },
              { seatNumber: 1, taken: "available" },
            ],
            [
              { seatNumber: 1, taken: "available" },
              { seatNumber: 1, taken: "available" },
            ],
          ],
          startTime: "2020-11-13T16:00:00.000Z",
        },
      ],
    },
  };
};

export const WithProvider = (props) => {
  const { state, dispatch } = getTestStore();
  return (
    <Theme>
      <AppContext.Provider
        value={{ state: props.defaultValue || state, dispatch }}
      >
        <Router>{props.children}</Router>
      </AppContext.Provider>
    </Theme>
  );
};
