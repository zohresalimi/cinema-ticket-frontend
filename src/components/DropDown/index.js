/* eslint-disable react/jsx-closing-tag-location */
/* eslint-disable react/jsx-indent */
/* eslint-disable react/jsx-no-comment-textnodes */
/* eslint-disable react/jsx-indent-props */
import React, { useState, useRef, useEffect } from "react";
import { useTranslation } from "react-i18next";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faAngleUp } from "@fortawesome/free-solid-svg-icons";
import { SUPPORTED_LOCALES } from "../../constants";
import { Wrapper, MenuLabel, ItemList, Button, CaratContainer } from "./style";

function DropDown() {
  const [isOpened, setIsOpened] = useState(false);
  const { i18n } = useTranslation();
  const [selectedOption, setSelectedOption] = useState(() => i18n.language);
  const [label, setLabel] = useState(
    () =>
      SUPPORTED_LOCALES.find((locale) => locale.code === selectedOption).name
  );

  const dropdownRef = useRef();
  useEffect(() => {
    const externalEventHandler = (e) => {
      if (!isOpened) return;

      const node = dropdownRef.current;

      if (node && node.contains(e.target)) {
        return;
      }

      setIsOpened(false);
    };

    if (isOpened) {
      document.addEventListener("click", externalEventHandler);
    } else {
      document.removeEventListener("click", externalEventHandler);
    }

    return () => {
      document.removeEventListener("click", externalEventHandler);
    };
  }, [isOpened]);

  const handleSelectedItem = (obj) => {
    setSelectedOption(obj.name);
    setLabel(obj.name);
    setIsOpened(!isOpened);
    i18n.changeLanguage(obj.code);
  };
  return (
    <Wrapper ref={dropdownRef}>
      <Button
        data-testid="toggle-dropdown"
        onClick={() => setIsOpened(!isOpened)}
      >
        <p>{selectedOption ? label : "svenska"}</p>
        <CaratContainer isOpen={isOpened}>
          <FontAwesomeIcon icon={faAngleUp} />
        </CaratContainer>
      </Button>
      <MenuLabel>
        {isOpened &&
          SUPPORTED_LOCALES.map((el) => (
            <ItemList key={el.code} onClick={() => handleSelectedItem(el)}>
              {el.name}
            </ItemList>
          ))}
      </MenuLabel>
    </Wrapper>
  );
}

export default DropDown;
