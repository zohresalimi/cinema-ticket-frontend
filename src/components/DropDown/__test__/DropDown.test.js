import * as React from "react";
import * as reactI18n from "react-i18next";
import { render, act, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import { getTestStore, WithProvider } from "../../../mockTestData/data";
import DropDown from "..";

function renderWrapper(props) {
  let component;

  act(() => {
    component = render(
      <WithProvider>
        <DropDown {...props} path="/" />
      </WithProvider>
    );
  });

  return component;
}

describe("DropDown Component Testing", () => {
  beforeAll(() => {
    const tMock = jest.fn((k) => k);
    const changeLanguageMock = jest.fn();
    const useTranslationMock = () => [
      tMock,
      {
        language: "sv",
        changeLanguage: changeLanguageMock,
      },
    ];

    jest
      .spyOn(reactI18n, "useTranslation")
      .mockImplementation(useTranslationMock);
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  test("take snapshot", async () => {
    const { container } = await renderWrapper();
    expect(container.firstChild).toMatchSnapshot();
  });

  test("page should be scrolled", async () => {
    const mockSetIsOpened = jest.fn(() => false);

    const { container, getByTestId } = await renderWrapper();
    fireEvent.click(getByTestId("toogle-dropdown"));
    expect(mockSetIsOpened).toHaveBeenCalled();
    expect(container.firstChild).toMatchSnapshot();
  });
  //   test("should play video", async () => {
  //     const setState = jest.fn();
  //     jest.spyOn(React, "useState").mockImplementation(() => [true, setState]);
  //     const movie = getTestStore().state.movies.premiered[0];

  //     const { container } = await renderWrapper({
  //       movie,
  //     });
  //     expect(container.firstChild).toMatchSnapshot();
  //   });
});
