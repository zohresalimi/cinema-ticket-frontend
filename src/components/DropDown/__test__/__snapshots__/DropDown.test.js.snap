// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown Component Testing take snapshot 1`] = `
<div
  style="outline: none;"
  tabindex="-1"
>
  <div
    class="style__Wrapper-sc-13kue4j-0 bvCERe"
  >
    <button
      class="style__Button-sc-13kue4j-3 dizWDM"
      data-testid="toogle-dropdown"
    >
      <p>
        Svenska
      </p>
      <div
        class="style__CaratContainer-sc-13kue4j-4 ljQyWE"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-angle-up fa-w-10 "
          data-icon="angle-up"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M177 159.7l136 136c9.4 9.4 9.4 24.6 0 33.9l-22.6 22.6c-9.4 9.4-24.6 9.4-33.9 0L160 255.9l-96.4 96.4c-9.4 9.4-24.6 9.4-33.9 0L7 329.7c-9.4-9.4-9.4-24.6 0-33.9l136-136c9.4-9.5 24.6-9.5 34-.1z"
            fill="currentColor"
          />
        </svg>
      </div>
    </button>
    <ul
      class="style__MenuLabel-sc-13kue4j-1 erqbDH"
    />
  </div>
</div>
`;
